### ISSUE TITLE
OS X MenuBar not working

### ISSUE THREAD

**Issue by omatuschin** [mgarin/weblaf#415]
I tried to add WebLaF to my existing Java project. However, I noticed that the menu bars are only appearing inside the JFrames, and not on top of the screen like they do when using the system L&F.
I am setting the apple.laf.useScreenMenuBar property directly after the startup of the application. Is this not supported by WebLaF, or am I doing something wrong?

Setting the MenuBarUI directly to AquaMenuBarUI (as described in https://github.com/mgarin/weblaf/issues/52) is causing an IllegalStateException: "Application.setDefaultMenuBar() only works with the Aqua Look and Feel".

**Comment by mgarin** [comment]
I'll be able to check this out tomorrow, but I'm pretty sure that the option should work with any L&F including WebLaF.

One more important thing - which WebLaF version did you use?

**Comment by omatuschin** [comment]
I'm using WebLaF 1.29. I think I also need to mention that I am using com.apple.eawt.Application.setDefaultMenuBar here, because my application needs to have a menu bar also when no windows are visible.

**Comment by mgarin** [comment]
Which build of v1.29 specifically are you using? - the one offered on the site or one of the latest `styling` branch builds from maven?

**Comment by omatuschin** [comment]
It is the one offered on the site.

**Comment by mgarin** [comment]
Okay, thanks :)
I will check this out as soon as I get my hands on a mac machine tomorrow.

**Comment by omatuschin** [comment]
Thanks for the fast response 👍

**Comment by mgarin** [comment]
Actually, I completely forgot to ask yesterday - you did mention WebLaF version, but which JDK & Mac OS X versions are you using? Those might also be quite important.

**Comment by omatuschin** [comment]
Sorry, I should have added that to my original description. Was probably too late at night... :-)
But I'll do so now...

**Comment by mgarin** [comment]
So I tried a few test examples and here are the results...

I've tried enabling native menu bar and as it was suggested by @Sciss in #52 - you have to setup a custom UI on the menu bar - `com.apple.laf.AquaMenuBarUI` one. There is no way to use native menu bar without using it because there is a hardcoded check associated with that UI within `com.apple.eawt._AppMenuBarHandler` class.

And that is not the only problem - if you would try to set that UI on the menu bar along with "apple.laf.useScreenMenuBar" property set to `true` you will get `UnsatisfiedLinkError` from a native utility method. That exception disappears only when you have Aqua L&F installed or at least initialized in runtime. If you look into Aqua L&F code you will find that it loads some native libraries upon initialization and there seem to be no way to load them separately - you will get multiple different exceptions if you try to load them outside of the Aqua L&F class.

That all leads to the only possible use case of the native menu with custom L&F:

1. Set property - `System.setProperty ( "apple.laf.useScreenMenuBar", "true" );`
2. Initialize Aqua L&F - `new AquaLookAndFeel ().initialize ();`
3. Setup Web L&F - `WebLookAndFeel.install ();`
4. Setup custom menu bar UI - `menuBar.setUI ( new AquaMenuBarUI () );`
5. Set frame menu bar - `frame.setJMenuBar ( menuBar );`

That menu bar will become native Mac OS X application menu bar. Any other menu bar set for other frames will also become native if they use `AquaMenuBarUI`, otherwise you will receive an exception upon adding menu bar to the frame.

Probably on the earlier JDK versions you could avoid step (2), but I'm not sure about that.

By the way, here is the environment I used for tests:

- OS X: Yosemite 10.10.5
- JDK: 8 update 102
- WebLaF: v1.29 pre-release build & v1.2.9 `styling` build

TL;DR 
Unfortunately there seem to be no reasonable way to setup native Mac OS X application menu without using the workaround described above. And sadly that has nothing to do with WebLaF or any other L&F - it is simply the way Mac OS X JDK utilities are written, pretty much forcing you to either use Aqua L&F or workarounds for custom L&Fs.

**Comment by mgarin** [comment]
I marked this issue as `JDK related` as this is pretty much related to the way proprietary Mac OS X JDK utilities (classes) work.

**Comment by omatuschin** [comment]
Thanks for your detailed response! I will try your suggestion setting the UI on the menubar again and let you know whether it worked. Might take some days however until I find the time for it :-)

**Comment by mgarin** [comment]
No problem, take your time. I will leave this issue to hang around in the tracker as it pretty much doesn't have a good solution so far (and I really doubt there will be anytime soon).

**Comment by myrlyn** [comment]
Menu bar not appearing on Windows 10 (Oracle JDK8) as well.  Moving down to weblaf 1.2.8 fixes the issue.

**Comment by mgarin** [comment]
@myrlyn I've tried this example:
```java
public final class MenuBarTest
{
    public static void main ( final String[] args )
    {
        SwingUtilities.invokeLater ( new Runnable ()
        {
            @Override
            public void run ()
            {
                WebLookAndFeel.install ();

                final JFrame frame = new JFrame ();

                final JMenuBar menuBar = new JMenuBar ();
                final JMenu menu = new JMenu ( "Menu 1" );
                menu.add ( new JMenuItem ( "Item 1" ) );
                menu.add ( new JMenuItem ( "Item 2" ) );
                menu.add ( new JMenuItem ( "Item 3" ) );
                menuBar.add ( menu );
                menuBar.add ( new JMenu ( "Menu 2" ) );
                menuBar.add ( new JMenu ( "Menu 3" ) );
                frame.setJMenuBar ( menuBar );

                final JLabel label = new JLabel ( "Content" );
                label.setBorder ( BorderFactory.createEmptyBorder ( 100, 150, 100, 150 ) );
                frame.add ( label );

                frame.pack ();
                frame.setLocationRelativeTo ( null );
                frame.setVisible ( true );
            }
        } );
    }
}
```
And it works for me just fine (Win 10 x64, JDK 8 update 211 x64).

Could you elaborate on your Windows/JDK versions?
Also could you provide any small code example that doesn't work correctly for you?

**Comment by mgarin** [comment]
@myrlyn 
It seems that there is a problem, but it's not related to OS X issue and it only appears on newer JDK versions (JDK9+) - there was an `updateUI ()` call added in `JRootPane.setMenuBar ( JMenuBar )` that actually corrupts existing menu bar layout.

I've added a separate issue #545 - I will be fixing it in v1.2.11 and will post an update once it is resolved.

**Comment by mgarin** [comment]
@myrlyn 
Menu bar disappearing should be fixed now for JDK 9+ (and earlier versions as well if it ever was the case).

**Comment by myrlyn** [comment]
Thanks!

**Comment by christianheilmann** [comment]
@kevinrushforth
@johanvos
I tested this fix and run in two issues still not fixed. 

1) If my application has a help-menu, the subentries of the help menu still stayed disabled (instead of enabled). A help-menu is rendered differently than the usual menus on macOS as it gets from macOS the search-bar attached. I think it recognizes the help-menu when the JMenu has the name "Help" and then renders it with the search bar in the menu.
All menu entries of the Help JMenu still stay disabled instead of enabled.
![SystemMenuBarHelpMenuEntries_Disabled](https://user-images.githubusercontent.com/24911444/101634739-66efe500-3a29-11eb-8167-22f63b7b5d55.png)

If I change by hand to another application and back to my java application, the menu entries of the help-menu still are not clickable and look like in the attached screenshot.

2) We have the case, that we need to disable the menu entries for a certain period of time while the application is running.
We use following code to do it:
Arrays.stream(menuBar.getSubElements()).forEach(menuElement -> {
            if (menuElement instanceof JMenu) {
                ((JMenu)menuElement).setEnabled(!disable);
            }
        });
        
After changing all JMenu to disable or enable, the displayed SystemMenuBar shows them not in actual state. I still have to change to another application and back to my java application, to see the menu entries in the right state.

I think changing the menubar at runtime leads to a problem, because as my understanding is, that you only initial at startup make the workaround to bring the application activated.

We have a mix of Swing and JavaFX in our application.

We try to strip down our application to give you a sample application for reproducing the following two errors.

**Comment by mgarin** [comment]
Before making a code sample - I would recommend testing this with Aqua L&F instead of WebLaF to see if the same issue occurs. If it does - this might actually be related to JDK implementation.

The thing is - WebLaF has no control over how that native menu works - "it just works". Swing handles the menu integration somewhere deep in it's native code. The only problem that was noticed is that it doesn't work if you don't have `AquaMenuBarUI` installed on your `JMenuBar` (mentioned in [this post](https://github.com/mgarin/weblaf/issues/415#issuecomment-254500109)).

**Comment by christianheilmann** [comment]
Thanks for your suggestion. I will try it.
